lib/cdo/contact_rollups/v2/pardot.rb

```diff
@@ -36,41 +47,34 @@ def self.retrieve_prospects(last_id, fields, limit = nil)
     raise ArgumentError.new("Missing value 'id' in fields argument") unless fields.include? 'id'
     total_results_retrieved = 0
 
+    # Limit the number of prospects retrieved in each API call if the overall limit is less than 200.
+    # @see http://developer.pardot.com/kb/api-version-4/prospects/#manipulating-the-result-set
+    limit_in_query = limit && limit < 200 ? limit : nil
+
     # Run repeated requests querying for prospects above our highest known Pardot ID.
+    # Stop when receiving no prospects or reaching the download limit.
     loop do
-      url = "#{PROSPECT_QUERY_URL}?id_greater_than=#{last_id}&fields=#{fields.join(',')}&sort_by=id"
-      url += "&limit=#{limit}" if limit
+      # Use bulk output format as recommended at http://developer.pardot.com/kb/bulk-data-pull/.
+      url = "#{PROSPECT_QUERY_URL}?output=bulk"
+      url += "&id_greater_than=#{last_id}&fields=#{fields.join(',')}&sort_by=id"
+      url += "&limit=#{limit_in_query}" if limit_in_query
       doc = post_with_auth_retry(url)
       raise_if_response_error(doc)
 
-      # Pardot returns the count total available prospects (not capped to 200),
-      # although the data for a max of 200 are contained in the response.
-      # The total prospects count changes in each loop iteration because the url we
-      # send to Pardot also changes.
-      # @see http://developer.pardot.com/kb/api-version-4/prospects/#xml-response-format
-      total_results = doc.xpath('/rsp/result/total_results').text.to_i
```

Comment:
One of the ways the bulk output optimizes for performance is to remove total_results calculation. This value is no longer returned. See:
http://developer.pardot.com/kb/bulk-data-pull/
http://developer.pardot.com/#changing-the-api-response-format